### PR TITLE
HIVE-24486 vertex merge

### PR DESCRIPTION
**Pull request by kgyrtkirk**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/ParallelEdgeFixer.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Lists;
+
+/**
+ * Inserts an extrea RS to avoid parallel edges.
+ *
+ * For mapjoins, semijoins its less costly to sometimes reshuffle the existing data - than computing it from scratch.
+ * Parallel edges are introduced by the {@link SharedWorkOptimizer} in case this fixer could patch them up.
+ */
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/ParallelEdgeFixer.java:55]: Could we add a small diagram (as others in SWO) representing the kind of plan shape this rule is fixing? In the past, this has helped quite a lot when we needed to revisit that code.

File: ql/src/test/results/clientpositive/llap/auto_join12.q.out [hunk 2]
```diff
@@ -128,6 +108,16 @@ STAGE PLANS:
                         Map-reduce partition columns: _col0 (type: string)
                         Statistics: Num rows: 166 Data size: 29548 Basic stats: COMPLETE Column stats: COMPLETE
                         value expressions: _col1 (type: string)
+                    Select Operator
+                      expressions: key (type: string)
+                      outputColumnNames: _col0
+                      Statistics: Num rows: 166 Data size: 14442 Basic stats: COMPLETE Column stats: COMPLETE
+                      Reduce Output Operator
+                        key expressions: _col0 (type: string)
```
Comment (jcamachor) [inline ql/src/test/results/clientpositive/llap/auto_join12.q.out:116]: Just a note (no action needed). It seems here we could have reused the same shuffle if it wasn't for the additional value column in one of them?

File: ql/src/test/results/clientpositive/llap/auto_join13.q.out [hunk 3]
```diff
@@ -91,42 +92,34 @@ STAGE PLANS:
             Map Operator Tree:
                 TableScan
                   alias: src
-                  filterExpr: (UDFToDouble(key) < 100.0D) (type: boolean)
+                  filterExpr: ((UDFToDouble(key) < 200.0D) or (UDFToDouble(key) < 100.0D)) (type: boolean)
```
Comment (jcamachor) [inline ql/src/test/results/clientpositive/llap/auto_join13.q.out:95]: Nice!

**Review by jcamachor — APPROVED** [review]

File: ql/src/test/results/clientpositive/llap/auto_join12.q.out (hunk 2 above)
Comment (kgyrtkirk) [reply ql/src/test/results/clientpositive/llap/auto_join12.q.out:116]: yes ..even if the tree is doing the same the rs is retained ;  we could probably improve on this later on...

**Pull request merged**